### PR TITLE
improve unit test performance

### PR DESCRIPTION
**Pull request by jackjohn7**
I updated the test logic to create one container per test suite. I created `withDb` abstraction where database access is protected behind a mutex. You must obtain access before you can run your test using the database. It integrates existing logic for the generator functions and typesafety.

### CODE REVIEW

File: src/lib/server/db/schema.ts [hunk 1]
```diff
@@ -23,6 +47,36 @@ export * from './users.sql';
...
+    userRoleTable,
+    sessionTable,
+    channelTagsTable,
+    postTagTable,
+    commentVoteTable,
+    postVoteTable,
+    postTable,
+    channelTable,
+    userTable,
+];
```
Comment (jackjohn7) [inline src/lib/server/db/schema.ts:78]: It'll likely be kinda easy for people to forget to add new tables to this list. It'd be sick to define a custom lint or a test to just validate that the length of this list is equal to the number of tables that we have.

File: src/lib/testing/utils.ts [hunk 2]
```diff
@@ -1,17 +1,42 @@
...
+                for (const table of tables) {
+                    await tx.delete(table);
+                }
+            });
+
+            const generated = generator ? await generator(db) : undefined;
+
+            await testFunc({ db, generated, expect });
+        });
+    });
```
Comment (funnyboy-roks) [inline src/lib/testing/utils.ts:41]: Would be nice to have access to the full object that is passed by vitest to the function in `test`.
i.e.,
```ts
test(name, async obj => {
    ...
    await testFunc({ db, generated, ...obj });
})
```

**Review by funnyboy-roks — CHANGES_REQUESTED** [review]
Looks good, just one note.

File: src/lib/server/db/schema.ts (hunk 1 above)
Comment (funnyboy-roks) [reply src/lib/server/db/schema.ts:78]: I think that would be nice, but definitely not important enough to block this PR.

Comment (jackjohn7) [reply src/lib/server/db/schema.ts]: I was able to avoid this entirely

Comment (funnyboy-roks) [reply src/lib/server/db/schema.ts]: Very snazzy

File: src/lib/testing/utils.ts (hunk 2 above)
Comment (jackjohn7) [reply src/lib/testing/utils.ts]: The type of obj seems to be weird such that `expect` isn't able to be destructured from it after the fact. Something breaks when I do this.

Comment (funnyboy-roks) [reply src/lib/testing/utils.ts]: I played with it a bit.  We can get this to work by doing
```ts
const newArgs = args as TestArgs<T>;
newArgs.db = db;
newArgs.generated = generated;

await testFunc(newArgs);
```
where `TestArgs` is
```ts
type TestArgs<T> = { db: DB; generated?: Awaited<T>; } & ExtendedContext<RunnerTestCase> & TestContext;
```

Comment (funnyboy-roks) [reply src/lib/testing/utils.ts]: My best guess is that `args` is a class with some funkiness for getting the `expect` or maybe it's using Proxies some how

File: src/lib/testing/utils.ts [hunk 3]
```diff
@@ -1,17 +1,47 @@
...
 
-    return { container: startedContainer, db, generated };
-};
+const tableNames: string[] = [];
+for (const tableName in db._.tableNamesMap) {
+    if (tableName != '') tableNames.push(tableName);
+}
 
 export const mustGenerate = <T>(generated: T | null) => {
     if (!generated) process.exit(1);
```
Comment (funnyboy-roks) [inline src/lib/testing/utils.ts:25]: Can you make this throw rather than just end the program?

File: src/lib/testing/utils.ts (hunk 2 above)
Comment (jackjohn7) [reply src/lib/testing/utils.ts]: Yeah it is kinda bizarre. What you suggested works though

File: src/lib/testing/utils.ts (hunk 3 above)
Comment (jackjohn7) [reply src/lib/testing/utils.ts]: Can indeed

File: src/lib/testing/utils.ts (hunk 2 above)
Comment (funnyboy-roks) [reply src/lib/testing/utils.ts]: Wonderful

File: src/lib/testing/utils.ts (hunk 3 above)
Comment (funnyboy-roks) [reply src/lib/testing/utils.ts]: Excellent

**Review by funnyboy-roks — APPROVED** [review]
I love it! They're so fast now!

**Pull request merged**